Sources/Core/Streams/Stream.swift

```diff
@@ -1,113 +1,34 @@
-            _ = try transformer.process(input)
-        }
-        
-        return transformer
-    }
+public protocol OutputStream: BaseStream {
+    associatedtype Output
+    typealias OutputHandler = (Output) -> ()
+    var outputStream: OutputHandler? { get set }
 }
 
-open class BasicStream<Output> : Stream {
-    /// Registers a closure that must be executed for every `Output` event
-    ///
-    /// - parameter closure: The closure to execute for each `Output` event
-    public func then(_ closure: @escaping ((Output) throws -> (Future<Void>))) {
-        listeners.append(closure)
-    }
-    
-    /// Writes output to all listeners
-    ///
-    /// - returns: A future that will be completed when all listeners have finished processing the output
-    @discardableResult
-    public func write(_ output: Output) throws -> Future<Void> {
-        let promise = Promise<Void>()
-        var counter = 0
-        
-        // listeners might be appended during execution
-        let expectedCount = listeners.count
-        
-        for listener in listeners {
-            try listener(output).then {
-                counter += 1
-                
-                if counter == expectedCount {
-                    _ = try? promise.complete(())
-                }
-            }
-        }
-        
-        return promise.future
-    }
-    
-    public init() {}
-    
-    /// Internal typealias used to define a cascading callback
-    fileprivate typealias Listener = ((Output) throws -> (Future<Void>))
-    
-    /// An internal array, used to keep track of all closures waiting for more data from this stream
-    fileprivate var listeners = [Listener]()
+public protocol BaseStream: class {
+    typealias ErrorHandler = (Error) -> ()
+    var errorStream: ErrorHandler? { get set }
 }
 
-/// A helper that allows you to transform streams
-open class StreamTransformer<From, To> : Stream {
-    /// The underlying stream helper
-    let stream = BasicStream<To>()
-    
-    /// Registers a closure that must be executed for every `Output` event
-    ///
-    /// - parameter closure: The closure to execute for each `Output` event
-    public func then(_ closure: @escaping ((To) throws -> (Future<Void>))) {
-        stream.then(closure)
+extension OutputStream {
+    public func consume(_ handler: @escaping OutputHandler) {
+        self.outputStream = handler
     }
-    
-    /// The input, that's being transformed
-    public typealias Input = From
-    
-    /// The resulting output
-    public typealias Output = To
-    
-    /// The transformer used to achieve the transformation from the input to the output
-    let transformer: ((Input) throws -> (Output?))
-    
-    /// Creates a new StreamTransformer using a closure
-    public init(using closure: @escaping ((Input) throws -> (Output?))) {
-        self.transformer = closure
+
+    public func stream<S: Stream>(to stream: S) -> S where S.Input == Self.Output {
```

Comment:
This is just syntactical, right? Since it's the same as Consume. Stream is a combination of `InputStream & OutputStream`. So consuming a Stream would call `consume` as if it were an `InputStream`.